File: 496-next-greater-element-i/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int>s;
        unordered_map<int,int>mpp;
        vector<int>res(nums1.size(),-1);
        for (int n:nums2){
            while (!s.empty() && s.top()<n){
                mpp[s.top()]=n;
                s.pop();
            }
            s.push(n);
        }
        for(int i=0;i<nums1.size();i++){
            if(mpp.find(nums1[i])!=mpp.end()){
                res[i]=mpp[nums1[i]];
            }
        }
        return res;
    }
};

File: 496-next-greater-element-i/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        res=[-1]*len(nums1)
        stack=[]
        map=defaultdict(int)
        for i in range(len(nums2)):
            while stack and stack[-1]<nums2[i]:
                map[stack.pop()]=nums2[i]
                
            stack.append(nums2[i])
        for i in range(len(nums1)):
            if nums1[i] in map:
                res[i]=map[nums1[i]]
        return res

            


        